pkg/streams/streams.go

```diff
@@ -0,0 +1,108 @@
+package streams
+
+import (
+	"context"
+	"sync"
+
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+// Stream is a stream of events
+type Stream struct {
+	// policy mask is a bitmap of policies that this stream is interested in
+	policyMask uint64
+	// events is a channel that is used to receive events from the stream
+	events chan trace.Event
+}
+
+// ReceiveEvents returns a channel that can be used to receive events from the stream
+func (s *Stream) ReceiveEvents() <-chan trace.Event {
+	return s.events
+}
+
+// Publish publishes an event to the stream,
+// but first check if this stream is interested in this event,
+// by checking the event's policy mask against the stream's policy mask.
+func (s *Stream) publish(ctx context.Context, event trace.Event) {
+	if s.shouldIgnorePolicy(event) {
+		return
+	}
+
+	// currently we are blocking if the channel is full,
+	// though I think we should instead drop events, because if you have
+	// two streams, one should not stop the other from receiving events,
+	// if they are not fast enough. But there also use cases where blocking
+	// might be favor, so th idea is to later make this configurable.
+	// TODO: allow this to be configurable (drop/block) (josedonizetti)
+	select {
+	case s.events <- event:
+	case <-ctx.Done():
+		return
+	}
+}
+
+// shouldIgnorePolicy checks if the stream should ignore the event
+func (s *Stream) shouldIgnorePolicy(event trace.Event) bool {
+	return s.policyMask&event.MatchedPoliciesUser == 0
+}
+
+// NewStream creates a new stream
+func NewStream(policyMask uint64, chanSize int) *Stream {
+	return &Stream{
+		policyMask: policyMask,
+		events:     make(chan trace.Event, chanSize),
+	}
+}
+
+// StreamManager manages streams
+type StreamsManager struct {
+	mutex       sync.Mutex
+	subscribers map[*Stream]struct{}
+}
+
+// NewStreamManager creates a new stream manager
+func NewStreamsManager() *StreamsManager {
+	return &StreamsManager{
+		mutex:       sync.Mutex{},
+		subscribers: make(map[*Stream]struct{}),
+	}
+}
+
+// Subscribe adds a stream to the manager
+func (sm *StreamsManager) Subscribe(stream *Stream) {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+	sm.subscribers[stream] = struct{}{}
+}
+
+// Unsubscribe removes a stream from the manager
+func (sm *StreamsManager) Unsubscribe(stream *Stream) {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	// check if stream is subscribed
+	if _, ok := sm.subscribers[stream]; ok {
+		delete(sm.subscribers, stream)
+		close(stream.events)
```

Comment:
Maybe close first and delete later? Wondering if another goroutine could have started using the StreamsManager (somehow) and continue with an already deleted reference even if closed. Just being overly cautious. 🤔

Reply:
I feel that encapsulation here is half baked (with manager closing the stream channel). I believe Unsubscribe should remove the stream from the manager, but what if others are still using it (I believe Geyslan thought the same). I would encapsulate entirely Stream and the StreamManager, subscbribe and unsubscribe a stream from the manager but without touching the stream. For the stream, if multiple threads are using it, you can even keep a "refcount" and only allow channel closure when refcount is 0, or anything like it (if ever needed).

Reply:
I don't think so! When considering channel semantics, we can always read from a closed channel, but we can never write to a closed channel, or the runtime panics. Considering the subscribers map is the state to know if we publish to a channel or not, we first stop publishing (writing), and then we notify readers, there won't be further data on the channel by closing it, the stream interface exposes a read only channel for this reason specially.

Reply:
If you think its good like this feel free to move on, I have no blockers.

Reply:
@rafaeldtinoco when I wrote the answer your comment wasn't appearing, let me read it, the first answer was for @geyslan comment about closing before.

Reply:
> I feel that encapsulation here is half baked (with manager closing the stream channel). I believe Unsubscribe should remove the stream from the manager, but what if others are still using it (I believe Geyslan thought the same). I would encapsulate entirely Stream and the StreamManager, subscbribe and unsubscribe a stream from the manager but without touching the stream. For the stream, if multiple threads are using it, you can even keep a "refcount" and only allow channel closure when refcount is 0, or anything like it (if ever needed).

@rafaeldtinoco But shouldn't the multiple goroutines (threads) know a channel won't be receiving data anymore? Or else the routine my keep running thinking events will be sent, but the publisher was removed, no data is coming there anymore. So closing works a signal, that once you drain the latest events from this steam either with `e,ok <-` or with `range`, you can stop the work, and close the routine.

Reply:
> I feel that encapsulation here is half baked (with manager closing the stream channel). I believe Unsubscribe should remove the stream from the manager, but what if others are still using it (I believe Geyslan thought the same).

Yeah I thought the same indeed, but as @josedonizetti remembered us, it seems that StreamsManager will be used in a very controlled area (not to be expanded in the codebase), so I agree that deleting it first will automatically remove it for writing which occurs only here:

![Screenshot from 2023-08-23 14-47-35](https://github.com/aquasecurity/tracee/assets/3372117/8329b9b5-f050-4d37-a715-5b14797e14c5)

Where we have a mutex control.

And closing it will make the channel only readable. So, yes I also have no blockers. 👍🏼

Reply:
Alright. Fair enough then.